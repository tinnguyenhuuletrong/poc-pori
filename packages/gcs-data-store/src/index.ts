export * from './lib/gcs-data-store';
