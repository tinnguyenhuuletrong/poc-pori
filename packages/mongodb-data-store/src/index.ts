export * from './lib/mongodb-data-store';
