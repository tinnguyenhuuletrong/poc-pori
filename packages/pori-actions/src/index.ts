import * as Input from './lib/input/pullData';
import * as DataView from './lib/dataView';

export * from './lib/basic';
export * from './lib/nftBodyPart';
export * from './lib/queryPoriApi';
export * from './lib/startStop';
export * from './lib/transformer/transformIdleGameEvent2Database';
export * from './lib/wallet/walletConnect';
export { Input, DataView };
