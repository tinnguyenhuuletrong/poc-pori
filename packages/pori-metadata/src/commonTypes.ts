import Web3 from 'web3';
import type { Account } from 'web3-core';
import type { Contract } from 'web3-eth-contract';
import type WalletConnect from '@walletconnect/client';
import type { WebsocketProvider, HttpProvider } from 'web3-core';
import { EventEmitter } from 'stream';
import { MongoClient } from 'mongodb';

export enum ENV {
  Staging = 'STAG',
  Prod = 'PROD',
  ProdPorichain = 'PROD_PORICHAIN',
}
export type SendMessageHandler = (msg: string) => Promise<any>;
export type EditMessageHandler = (
  lastMsginfo: any,
  msg: string
) => Promise<any>;
export type CustomEstGasprice = (ctx: Context) => Promise<string>;

export type Context = {
  contract: Contract;
  web3: Web3;
  walletAcc?: Account;
  provider: WebsocketProvider | HttpProvider;
  walletConnectChannel?: WalletConnect;
  env: ENV;
  emiter?: EventEmitter;
  mongoClient?: MongoClient;
  playerAddress?: string;

  ui: {
    writeMessage: SendMessageHandler;
    editMessage: EditMessageHandler;
  };
  setting: {
    gasFactor: number;
    safeGweith: number;
    autoPlayMicroDelayMs: number;
  };
  custom: {
    estimageGas?: CustomEstGasprice;
  };
};

export type DnaBodyPartInfo = {
  d_head_class: number;
  d_head_type: number;
  r1_head_class: number;
  r1_head_type: number;
  r2_head_class: number;
  r2_head_type: number;
  r3_head_class: number;
  r3_head_type: number;
  d_face_class: number;
  d_face_type: number;
  r1_face_class: number;
  r1_face_type: number;
  r2_face_class: number;
  r2_face_type: number;
  r3_face_class: number;
  r3_face_type: number;
  d_body_class: number;
  d_body_type: number;
  r1_body_class: number;
  r1_body_type: number;
  r2_body_class: number;
  r2_body_type: number;
  r3_body_class: number;
  r3_body_type: number;
  d_arm_class: number;
  d_arm_type: number;
  r1_arm_class: number;
  r1_arm_type: number;
  r2_arm_class: number;
  r2_arm_type: number;
  r3_arm_class: number;
  r3_arm_type: number;
  d_accessory_class: number;
  d_accessory_type: number;
  r1_accessory_class: number;
  r1_accessory_type: number;
  r2_accessory_class: number;
  r2_accessory_type: number;
  r3_accessory_class: number;
  r3_accessory_type: number;
  d_leg_class: number;
  d_leg_type: number;
  r1_leg_class: number;
  r1_leg_type: number;
  r2_leg_class: number;
  r2_leg_type: number;
  r3_leg_class: number;
  r3_leg_type: number;
  d_tail_class: number;
  d_tail_type: number;
  r1_tail_class: number;
  r1_tail_type: number;
  r2_tail_class: number;
  r2_tail_type: number;
  r3_tail_class: number;
  r3_tail_type: number;
  legendary: number;
  reserved: number;
  l_head: number;
  l_face: number;
  l_body: number;
  l_arm: number;
  l_accessory: number;
  l_leg: number;
  l_tail: number;
  species_type: number;
  species_class: string;
  d_head_type_name: string;
  d_face_type_name: string;
  d_body_type_name: string;
  d_arm_type_name: string;
  d_accessory_type_name: string;
  d_leg_type_name: string;
  d_tail_type_name: string;
  r1_head_type_name: string;
  r1_face_type_name: string;
  r1_body_type_name: string;
  r1_arm_type_name: string;
  r1_accessory_type_name: string;
  r1_leg_type_name: string;
  r1_tail_type_name: string;
  r2_head_type_name: string;
  r2_face_type_name: string;
  r2_body_type_name: string;
  r2_arm_type_name: string;
  r2_accessory_type_name: string;
  r2_leg_type_name: string;
  r2_tail_type_name: string;
  r3_head_type_name: string;
  r3_face_type_name: string;
  r3_body_type_name: string;
  r3_arm_type_name: string;
  r3_accessory_type_name: string;
  r3_leg_type_name: string;
  r3_tail_type_name: string;
  purity: number;
};

export type NftInfo = {
  tokenId: number;
  type: number;
  name: string;
  dna: string;
  status: number;
  poriClass: number;
  numOfBreeds: number;
  maxOfBreeds: number;
  legend: number;
  purity: number;
  birthDate: string;
  ownerAddress: string;
  stage: number;
  health: number;
  speed: number;
  physicalAttack: number;
  physicalDefend: number;
  critical: number;
  magicAttack: number;
  magicDefend: number;
  minePower: number;
  helpPower: number;
  autoRenew: boolean;
  chapter1ExpiredAt: number | null;
  createdAt: string;
  updatedAt: string;
  ownerInfo: {
    ownerId: string;
    fullName: string;
    publicAddress: string;
  };
};

export type AdventureInfo = {
  mineId: number;

  state?: string;

  isFarmer: boolean;
  farmerAddress?: string;
  startTime?: Date;
  farmerEndTime: Date;

  isSupporter?: boolean;
  supporterAddress?: string;
  supporterEndTime?: Date;
  winnerAddress?: string;

  farmerPories?: number[];
  farmerRewardLevel?: number[];
  farmerSlots?: number[];

  supporterPories?: number[];
  supporterRewardLevel?: number[];
  supporterSlots?: number[];
};

export type AdventureStatsGroupByDay = {
  unixDay: number;
  timestamp: Date;
  finishedMineIds: number[];
  totalRigy: number;
  totalRiken: number;
};

export type AdventureInfoEx = Omit<
  AdventureInfo,
  'startTime' | 'farmerEndTime' | 'supporterEndTime'
> & {
  link: string;
  canCollect?: boolean;
  hasBigReward: boolean;
  startTime?: string;
  farmerEndTime?: string;
  supporterEndTime?: string;
  farmerAtkTime?: string;
  supporterAtkTime?: string;
  blockedTo: Date;
  atkAt: Date;
  powers: Record<string, number>;
};

export type PromiseReturnType<T> = T extends Promise<infer Return> ? Return : T;
