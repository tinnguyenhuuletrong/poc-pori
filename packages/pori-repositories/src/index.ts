import Realm, { ConfigurationWithoutSync } from 'realm';
import { Schemas as IdleGameSchemas } from './lib/idleGames/schema';
import {
  IdleGameSCMetadataRepo,
  IdleGameSCMetadataDataModel,
  IdleGameSCEventRepo,
  IdleGameSCEventDataModel,
  PlayerRepo,
  PlayerDataModel,
  PoriRepo,
  PoriDataModel,
} from './lib/idleGames/schema';

export async function openRepo(opt: ConfigurationWithoutSync) {
  const schemas = opt.schema ?? [];
  const ins = await Realm.open({
    ...opt,
    schemaVersion: 2,
    schema: [...schemas, ...IdleGameSchemas],
  });
  return ins;
}

export {
  IdleGameSCMetadataDataModel,
  IdleGameSCEventDataModel,
  IdleGameSCMetadataRepo,
  IdleGameSCEventRepo,
  PlayerRepo,
  PlayerDataModel,
  PoriRepo,
  PoriDataModel,
};
