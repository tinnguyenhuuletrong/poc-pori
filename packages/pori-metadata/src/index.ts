import { Context, CustomEstGasprice, ENV } from './commonTypes';
import * as stagConfig from './lib/sta-poriverse_info';
import * as prodConfig from './lib/prod-poriverse_info';
import * as prodPoriChainConfig from './lib/prod-porichain-poriverse_info';
import { mean } from 'lodash';

export const TEN_POWER_10_BN = BigInt(10 ** 18);
export const TURN_DURATION_SEC = 1800;

export function getWeb3NodeUri(env: ENV) {
  const key = `NODE_URI_${env}`.toUpperCase();
  return process.env[key] as string;
}

export function getWeb3NodeUriHttp(env: ENV) {
  const key = `NODE_URI_${env}_HTTP`.toUpperCase();
  return process.env[key] as string;
}

export function getWeb3NodeUriPolygonHttp() {
  const key = `NODE_URI_${ENV.Prod}_HTTP`.toUpperCase();
  return process.env[key] as string;
}

export function getAPILink(env: ENV) {
  if (env === ENV.Staging) return stagConfig.gameInfo.m.app.apiUrl;
  else if (env === ENV.ProdPorichain)
    return prodPoriChainConfig.gameInfo.m.app.apiUrl;
  else return prodConfig.gameInfo.m.app.apiUrl;
}

export function getKyberSwapFactoryAddress(env: ENV) {
  // https://docs.kyberswap.com/developer-guides/kyberswap-addresses#mainnet

  if (env === ENV.Staging) return '0x7900309d0b1c8D3d665Ae40e712E8ba4FC4F5453';
  return '0x5f1fe642060b5b9658c15721ea22e982643c095c';
}

export function getRIGYTokenInfoOnPolygon() {
  const tokenConfig = prodConfig.gameInfo.m.app;
  return {
    symbol: tokenConfig.token.inGameSymbol,
    tokenAddress: tokenConfig.token.inGameAddress,
    decimal: tokenConfig.token.inGameDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getRIGYTokenInfo(env: ENV) {
  let tokenConfig = prodConfig.gameInfo.m.app;
  if (env === ENV.Staging) {
    tokenConfig = stagConfig.gameInfo.m.app;
  } else if (env === ENV.ProdPorichain)
    tokenConfig = prodPoriChainConfig.gameInfo.m.app;

  return {
    symbol: tokenConfig.token.inGameSymbol,
    tokenAddress: tokenConfig.token.inGameAddress,
    decimal: tokenConfig.token.inGameDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getRIKENTokenInfoOnPolygon() {
  const tokenConfig = prodConfig.gameInfo.m.app;
  return {
    symbol: tokenConfig.token.nativeSymbol,
    tokenAddress: tokenConfig.token.nativeAddress,
    decimal: tokenConfig.token.nativeDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getRIKENTokenInfo(env: ENV) {
  let tokenConfig = prodConfig.gameInfo.m.app;
  if (env === ENV.Staging) {
    tokenConfig = stagConfig.gameInfo.m.app;
  } else if (env === ENV.ProdPorichain)
    tokenConfig = prodPoriChainConfig.gameInfo.m.app;

  return {
    symbol: tokenConfig.token.nativeSymbol,
    tokenAddress: tokenConfig.token.nativeAddress,
    decimal: tokenConfig.token.nativeDecimal,
    chainId: tokenConfig.rpcMetamask.chainId,
  };
}

export function getAdventureBaseLink(env: ENV) {
  if (env === ENV.Staging) return stagConfig.gameInfo.m.app.urlAdventure;
  else if (env === ENV.ProdPorichain)
    return prodPoriChainConfig.gameInfo.m.app.urlAdventure;
  return prodConfig.gameInfo.m.app.urlAdventure;
}

export function getMarketplayBaseLink(env: ENV) {
  if (env === ENV.Staging) return stagConfig.gameInfo.m.app.urlMarketplace;
  else if (env === ENV.ProdPorichain)
    return prodPoriChainConfig.gameInfo.m.app.urlMarketplace;
  return prodConfig.gameInfo.m.app.urlMarketplace;
}

export function getIdleGameAddressSC(env: ENV) {
  if (env === ENV.Staging) {
    return {
      abi: stagConfig.ABI_IDLE,
      address: stagConfig.gameInfo.m.app.contractAddress.idleGameAddress,
      createdBlock: stagConfig.gameInfo.m.app.scCreatedBlock.idle,
    };
  } else if (env === ENV.ProdPorichain)
    return {
      abi: prodPoriChainConfig.ABI_IDLE,
      address:
        prodPoriChainConfig.gameInfo.m.app.contractAddress.idleGameAddress,
      createdBlock: prodPoriChainConfig.gameInfo.m.app.scCreatedBlock.idle,
    };
  return {
    abi: prodConfig.ABI_IDLE,
    address: prodConfig.gameInfo.m.app.contractAddress.idleGameAddress,
    createdBlock: prodConfig.gameInfo.m.app.scCreatedBlock.idle,
  };
}

export function calculateMineTurnTime(startTime: Date) {
  const farmerAtkStartAt = new Date(
    startTime.valueOf() + TURN_DURATION_SEC * 1000
  );
  const supporterAtkStartAt = new Date(
    startTime.valueOf() + TURN_DURATION_SEC * 2 * 1000
  );
  return {
    farmerAtkStartAt,
    supporterAtkStartAt,
  };
}

export function getContextSetting(env: ENV) {
  // getGas porichain return as 500. But avg fee on chain around 4100
  const gasFactor = 1;
  const safeGweith = env === ENV.ProdPorichain ? 5000 : 80;
  const autoPlayMicroDelayMs = env === ENV.ProdPorichain ? 10000 : 3000;

  // custom estimage gas function base on pending tx for porichain
  let estimageGas: CustomEstGasprice | undefined;
  if (env === ENV.ProdPorichain) {
    estimageGas = async (ctx: Context) => {
      const pendingTx = await ctx.web3.eth.getPendingTransactions();
      const avgGas = mean(
        pendingTx
          .filter((itm) => +itm.gasPrice > 0)
          .map((itm) => {
            return +itm.gasPrice;
          })
      );
      return avgGas.toString();
    };
  }

  return {
    setting: { gasFactor, safeGweith, autoPlayMicroDelayMs },
    custom: {
      estimageGas,
    },
  };
}

export * from './commonTypes';

export * as IdleGameSc from './lib/idleGameSc';
export * from './lib/idleGameSc/type.idleGame';
