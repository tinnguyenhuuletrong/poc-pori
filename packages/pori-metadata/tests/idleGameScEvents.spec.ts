import type { EventData } from 'web3-eth-contract';
import { IdleGameSc } from '../src';

const SAMPLE_DATA = [
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25786299,
    transactionHash:
      '0x97ea12da694c241a29ea893676a6f964b29d9a4d92e38ba152e06003aa38481c',
    transactionIndex: 40,
    blockHash:
      '0xeccb1b6fae245adba3b8ed8a3fe831e45d8bc63316a3654ae84143060483648b',
    logIndex: 241,
    removed: false,
    id: 'log_fdab3d28',
    returnValues: {
      '0': '6',
      '1': '0xf9209B6F49BB9fD73422BA834f4cD444aE7ceacE',
      '2': '1649006285',
      '3': ['24', '38', '37'],
      '4': ['2', '3', '7'],
      '5': ['3', '1', '4'],
      '6': '1649007785',
      mineId: '6',
      farmer: '0xf9209B6F49BB9fD73422BA834f4cD444aE7ceacE',
      startTime: '1649006285',
      porians: ['24', '38', '37'],
      indexes: ['2', '3', '7'],
      rewardLevels: ['3', '1', '4'],
      blockedTime: '1649007785',
    },
    event: 'AdventureStarted',
    signature:
      '0xd498194e39f0d0d9426ee530bd16b2182b34d07d968365c8fdcaf73c5a6d0ac5',
    raw: {
      data: '0x000000000000000000000000000000000000000000000000000000000000001800000000000000000000000000000000000000000000000000000000000000260000000000000000000000000000000000000000000000000000000000000025000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000007000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000006249dca9',
      topics: [
        '0xd498194e39f0d0d9426ee530bd16b2182b34d07d968365c8fdcaf73c5a6d0ac5',
        '0x0000000000000000000000000000000000000000000000000000000000000006',
        '0x000000000000000000000000f9209b6f49bb9fd73422ba834f4cd444ae7ceace',
        '0x000000000000000000000000000000000000000000000000000000006249d6cd',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25786571,
    transactionHash:
      '0x9f8a3470f045d389a21948d5aa72231b9da5fa4f32905e80eb23dd967fe7f19a',
    transactionIndex: 45,
    blockHash:
      '0x8c2fa6e193525c76fbfab57899178ca4404b562c9ee6a3dc9d45915743e286aa',
    logIndex: 305,
    removed: false,
    id: 'log_c0dbaff9',
    returnValues: {
      '0': '6',
      '1': '0x1F8A3F6083F90f5FeF56AC3D3F22822C41C9488a',
      '2': '0',
      '3': '43600000000000000000',
      '4': '4360000000000000000',
      '5': '82400000000000000000',
      '6': '8240000000000000000',
      mineId: '6',
      winner: '0x1F8A3F6083F90f5FeF56AC3D3F22822C41C9488a',
      fragments: '0',
      farmerReward1: '43600000000000000000',
      farmerReward2: '4360000000000000000',
      helperReward1: '82400000000000000000',
      helperReward2: '8240000000000000000',
    },
    event: 'AdventureFinished',
    signature:
      '0xacba132576685783d626ee7ff7486ac6cf8580b51ca1ef49ee36edb6303ac735',
    raw: {
      data: '0x0000000000000000000000000000000000000000000000025d1250ff4f0800000000000000000000000000000000000000000000000000003c81d4e654b4000000000000000000000000000000000000000000000000000477879b6d14300000000000000000000000000000000000000000000000000000725a5c57b5380000',
      topics: [
        '0xacba132576685783d626ee7ff7486ac6cf8580b51ca1ef49ee36edb6303ac735',
        '0x0000000000000000000000000000000000000000000000000000000000000006',
        '0x0000000000000000000000001f8a3f6083f90f5fef56ac3d3f22822c41c9488a',
        '0x0000000000000000000000000000000000000000000000000000000000000000',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25799104,
    transactionHash:
      '0x4f91872215a9822cd300c0d0f11dd28083aa2383734797355b3dcce2048171cf',
    transactionIndex: 32,
    blockHash:
      '0xa8454d4a6d014b46784df8272d2b9b3656927b87f7ce34062018e67b11d569ed',
    logIndex: 195,
    removed: false,
    id: 'log_7c14a03f',
    returnValues: {
      '0': '17',
      '1': '27',
      '2': '7',
      '3': '4',
      '4': '1649088619',
      mineId: '17',
      porian: '27',
      index: '7',
      rewardLevel: '4',
      blockedTime: '1649088619',
    },
    event: 'AdventureFortified',
    signature:
      '0x71b4d764a280d810a1907567eb53bcf7ebe267f0d94d40d2a5f20009e7b73569',
    raw: {
      data: '0x000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000624b186b',
      topics: [
        '0x71b4d764a280d810a1907567eb53bcf7ebe267f0d94d40d2a5f20009e7b73569',
        '0x0000000000000000000000000000000000000000000000000000000000000011',
        '0x000000000000000000000000000000000000000000000000000000000000001b',
        '0x0000000000000000000000000000000000000000000000000000000000000007',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25799448,
    transactionHash:
      '0x2aebec81d7a633a90959194ecdc861824a392693e263baf5c6d8fe6961fc3fba',
    transactionIndex: 26,
    blockHash:
      '0x28dea73b17c6869e8c4a85b371530c8a516e80283af392356913008322867fac',
    logIndex: 178,
    removed: false,
    id: 'log_3995e08c',
    returnValues: {
      '0': '18',
      '1': '0xf9209B6F49BB9fD73422BA834f4cD444aE7ceacE',
      '2': ['30', '34', '27'],
      '3': ['7', '8', '9'],
      '4': ['4', '3', '3'],
      '5': '1649090860',
      mineId: '18',
      helper: '0xf9209B6F49BB9fD73422BA834f4cD444aE7ceacE',
      porians: ['30', '34', '27'],
      indexes: ['7', '8', '9'],
      rewardLevels: ['4', '3', '3'],
      blockedTime: '1649090860',
    },
    event: 'AdventureSupported1',
    signature:
      '0xbe2e74f68284a904ef29e10f3e20b2c9bb540481fb9903c1aead3e26cc56f8b1',
    raw: {
      data: '0x000000000000000000000000000000000000000000000000000000000000001e0000000000000000000000000000000000000000000000000000000000000022000000000000000000000000000000000000000000000000000000000000001b00000000000000000000000000000000000000000000000000000000000000070000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000900000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000000000000000000000000000624b212c',
      topics: [
        '0xbe2e74f68284a904ef29e10f3e20b2c9bb540481fb9903c1aead3e26cc56f8b1',
        '0x0000000000000000000000000000000000000000000000000000000000000012',
        '0x000000000000000000000000f9209b6f49bb9fd73422ba834f4cd444ae7ceace',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25799513,
    transactionHash:
      '0xeea65ae96f82956d8cd0b1b004f120f72c2ca010129663ad5afbf8667ea934fc',
    transactionIndex: 94,
    blockHash:
      '0x9d9b73065a281291cc592debf45630528e5f9a7912d634107a0ead29a7ae5098',
    logIndex: 392,
    removed: false,
    id: 'log_f7830dec',
    returnValues: {
      '0': '18',
      '1': '350',
      '2': '7',
      '3': '4',
      '4': '1649090860',
      mineId: '18',
      porian: '350',
      index: '7',
      rewardLevel: '4',
      blockedTime: '1649090860',
    },
    event: 'AdventureSupported2',
    signature:
      '0x6dbf0858232497280bfdf35e37adf2002793779d07d4f48947b17509c71dd41c',
    raw: {
      data: '0x000000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000624b212c',
      topics: [
        '0x6dbf0858232497280bfdf35e37adf2002793779d07d4f48947b17509c71dd41c',
        '0x0000000000000000000000000000000000000000000000000000000000000012',
        '0x000000000000000000000000000000000000000000000000000000000000015e',
        '0x0000000000000000000000000000000000000000000000000000000000000007',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25799529,
    transactionHash:
      '0xa0e86ebf2f6e0a484a0428db7f5f6f9e0cbda347eb4f06b1cfa21aa7afb83408',
    transactionIndex: 43,
    blockHash:
      '0x5aada42876f872d35ff03ac580b788f94d313a0b5cd0d1737ef64595213e7563',
    logIndex: 130,
    removed: false,
    id: 'log_9885030e',
    returnValues: {
      '0': '0x8d1A4b353d80f31757CC2264f6a163c410B2fBF2',
      '1': '102',
      '2': '7955107200',
      from: '0x8d1A4b353d80f31757CC2264f6a163c410B2fBF2',
      porian: '102',
      expiredAt: '7955107200',
    },
    event: 'PorianDeposited',
    signature:
      '0xc42131d410ea79f1eafecd549df9a392681974b7f4d9f4d78c216ea12b6766e8',
    raw: {
      data: '0x',
      topics: [
        '0xc42131d410ea79f1eafecd549df9a392681974b7f4d9f4d78c216ea12b6766e8',
        '0x0000000000000000000000008d1a4b353d80f31757cc2264f6a163c410b2fbf2',
        '0x0000000000000000000000000000000000000000000000000000000000000066',
        '0x00000000000000000000000000000000000000000000000000000001da294d80',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25836477,
    transactionHash:
      '0xd2162717f59565bbd9f6c3b3a385c861a96bac487a7e4eeca3d5ef899542defa',
    transactionIndex: 24,
    blockHash:
      '0x4c58272bf18ee4f0dc7831ea0bde32ea2d79abaedf99bfa0344b24643495fcfe',
    logIndex: 109,
    removed: false,
    id: 'log_75c6a6ff',
    returnValues: {
      '0': '0x0a674eB963646bE42F9F071B8F5Cd79Bb744bD32',
      '1': '9116',
      to: '0x0a674eB963646bE42F9F071B8F5Cd79Bb744bD32',
      porian: '9116',
    },
    event: 'PorianWithdrawed',
    signature:
      '0x44402a61584354899786311a4f0c7bf924b31db70b5ebef891d88cee08156ed5',
    raw: {
      data: '0x',
      topics: [
        '0x44402a61584354899786311a4f0c7bf924b31db70b5ebef891d88cee08156ed5',
        '0x0000000000000000000000000a674eb963646be42f9f071b8f5cd79bb744bd32',
        '0x000000000000000000000000000000000000000000000000000000000000239c',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25777544,
    transactionHash:
      '0x541e5087cb1cd02e3991abcd1e43e0aeb0b158e4400292c58d059976412d32d2',
    transactionIndex: 29,
    blockHash:
      '0x45a4f28dd6a9f9398e5bab5c7dd6a6ab4c1d0b3a754ad20906546c99e6d3cc5d',
    logIndex: 156,
    removed: false,
    id: 'log_6165203a',
    returnValues: {
      '0': '0x0000000000000000000000000000000000000000',
      '1': '0xf430aD5c000889aB42a68790A26A0Ee6c460D240',
      previousOwner: '0x0000000000000000000000000000000000000000',
      newOwner: '0xf430aD5c000889aB42a68790A26A0Ee6c460D240',
    },
    event: 'OwnershipTransferred',
    signature:
      '0x8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0',
    raw: {
      data: '0x',
      topics: [
        '0x8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e0',
        '0x0000000000000000000000000000000000000000000000000000000000000000',
        '0x000000000000000000000000f430ad5c000889ab42a68790a26a0ee6c460d240',
      ],
    },
  },
  {
    address: '0x70d66d12c424Db41bf4E501A3FBbF5D4a779E66b',
    blockNumber: 25777545,
    transactionHash:
      '0xf59a96c14408acef73996d4e76278bf5830a60fa0139188d66f15788a2270891',
    transactionIndex: 22,
    blockHash:
      '0xd1735ba174f43224de4695edc17ea3302fc075fec36a569bd348536ca29e9b93',
    logIndex: 90,
    removed: false,
    id: 'log_d661f256',
    returnValues: {
      '0': '1500',
      '1': '300',
      adventureDuration: '1500',
      turnDuration: '300',
    },
    event: 'GameDurationChanged',
    signature:
      '0xc62be04bfb76e5e364578771d33bb80ebbea7219b67ba068fb4bbdaf83e4a3c0',
    raw: {
      data: '0x',
      topics: [
        '0xc62be04bfb76e5e364578771d33bb80ebbea7219b67ba068fb4bbdaf83e4a3c0',
        '0x00000000000000000000000000000000000000000000000000000000000005dc',
        '0x000000000000000000000000000000000000000000000000000000000000012c',
      ],
    },
  },
  {
    address: '0xd678f37434d06Fc3Eb43c11024998FAaFd411881',
    blockHash:
      '0x55d5f83157669d423cfce6d022c59e5b969c12252cf0e75d13e4f2fb478fb115',
    blockNumber: 696069,
    logIndex: 3,
    removed: false,
    transactionHash:
      '0x146f66a90cdea6225aa8b5d37a85cc44c48ab24117fae7cb948cf63140c472f0',
    transactionIndex: 1,
    transactionLogIndex: '0x1',
    id: 'log_efd5427d',
    returnValues: {},
    event: undefined,
    signature: null,
    raw: {
      data: '0x000000000000000000000000000000000000000000000000000000000000ca78000000000000000000000000df218bd4414e0b1d581bddf64498abba8cce0eca000000000000000000000000000000000000000000000000000000000001889c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000090000000000000000000000000000000000000000000000000000000000000007',
      topics: [
        '0x6bdac8de130455f3dbd97b2916ff758c9cb534e9770af694954b63b76169d728',
      ],
    },
  } as any,
];

test('parse idleGameScEvent', () => {
  const res = SAMPLE_DATA.map((itm) => IdleGameSc.parseIdleGameScEvent(itm));
  expect(res).toMatchSnapshot();
});
