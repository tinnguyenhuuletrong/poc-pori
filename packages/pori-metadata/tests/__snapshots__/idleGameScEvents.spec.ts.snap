// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse idleGameScEvent 1`] = `
Array [
  Object {
    "blockNo": 25786299,
    "data": Object {
      "blockedTime": 1649007785,
      "farmer": "0xf9209B6F49BB9fD73422BA834f4cD444aE7ceacE",
      "indexes": Array [
        2,
        3,
        7,
      ],
      "mineId": 6,
      "porians": Array [
        24,
        38,
        37,
      ],
      "rewardLevels": Array [
        3,
        1,
        4,
      ],
      "startTime": 1649006285,
    },
    "txHash": "0x97ea12da694c241a29ea893676a6f964b29d9a4d92e38ba152e06003aa38481c",
    "type": "AdventureStarted",
  },
  Object {
    "blockNo": 25786571,
    "data": Object {
      "farmerReward1": Object {
        "$numberDecimal": "43600000000000000000",
      },
      "farmerReward2": Object {
        "$numberDecimal": "4360000000000000000",
      },
      "fragments": 0,
      "helperReward1": Object {
        "$numberDecimal": "82400000000000000000",
      },
      "helperReward2": Object {
        "$numberDecimal": "8240000000000000000",
      },
      "mineId": 6,
      "winner": "0x1F8A3F6083F90f5FeF56AC3D3F22822C41C9488a",
    },
    "txHash": "0x9f8a3470f045d389a21948d5aa72231b9da5fa4f32905e80eb23dd967fe7f19a",
    "type": "AdventureFinished",
  },
  Object {
    "blockNo": 25799104,
    "data": Object {
      "blockedTime": 1649088619,
      "index": 7,
      "mineId": 17,
      "porian": 27,
      "rewardLevel": 4,
    },
    "txHash": "0x4f91872215a9822cd300c0d0f11dd28083aa2383734797355b3dcce2048171cf",
    "type": "AdventureFortified",
  },
  Object {
    "blockNo": 25799448,
    "data": Object {
      "blockedTime": 1649090860,
      "helper": "0xf9209B6F49BB9fD73422BA834f4cD444aE7ceacE",
      "indexes": Array [
        7,
        8,
        9,
      ],
      "mineId": 18,
      "porians": Array [
        30,
        34,
        27,
      ],
      "rewardLevels": Array [
        4,
        3,
        3,
      ],
    },
    "txHash": "0x2aebec81d7a633a90959194ecdc861824a392693e263baf5c6d8fe6961fc3fba",
    "type": "AdventureSupported1",
  },
  Object {
    "blockNo": 25799513,
    "data": Object {
      "blockedTime": 1649090860,
      "index": 7,
      "mineId": 18,
      "porian": 350,
      "rewardLevel": 4,
    },
    "txHash": "0xeea65ae96f82956d8cd0b1b004f120f72c2ca010129663ad5afbf8667ea934fc",
    "type": "AdventureSupported2",
  },
  Object {
    "blockNo": 25799529,
    "data": Object {
      "expiredAt": 7955107200,
      "from": "0x8d1A4b353d80f31757CC2264f6a163c410B2fBF2",
      "porian": 102,
    },
    "txHash": "0xa0e86ebf2f6e0a484a0428db7f5f6f9e0cbda347eb4f06b1cfa21aa7afb83408",
    "type": "PorianDeposited",
  },
  Object {
    "blockNo": 25836477,
    "data": Object {
      "porian": 9116,
      "to": "0x0a674eB963646bE42F9F071B8F5Cd79Bb744bD32",
    },
    "txHash": "0xd2162717f59565bbd9f6c3b3a385c861a96bac487a7e4eeca3d5ef899542defa",
    "type": "PorianWithdrawed",
  },
  Object {
    "blockNo": 25777544,
    "data": undefined,
    "txHash": "0x541e5087cb1cd02e3991abcd1e43e0aeb0b158e4400292c58d059976412d32d2",
    "type": "OwnershipTransferred",
  },
  Object {
    "blockNo": 25777545,
    "data": Object {
      "adventureDuration": 1500,
      "turnDuration": 300,
    },
    "txHash": "0xf59a96c14408acef73996d4e76278bf5830a60fa0139188d66f15788a2270891",
    "type": "GameDurationChanged",
  },
  Object {
    "blockNo": 696069,
    "data": Object {
      "farmer": "0xdF218Bd4414E0B1D581BDdF64498ABBa8cCe0EcA",
      "from": "9",
      "helper": undefined,
      "mineId": 51832,
      "porian": 100508,
      "to": "7",
    },
    "txHash": "0x146f66a90cdea6225aa8b5d37a85cc44c48ab24117fae7cb948cf63140c472f0",
    "type": "SBattleSwapped",
  },
]
`;
